src/copy.rs

```diff
@@ -0,0 +1,430 @@
+    /// copy operation no longer wrote data.
+    ///
+    /// This does not include any padding at the end necessary to maintain
+    /// alignment requirements.
+    ///
+    /// Unless you really know what you're doing, you *likely* want to use `end_offset_padded` instead.
+    pub copy_end_offset: usize,
+
+    /// The offset from the start of the allocation, in bytes, at which the
+    /// copy operation no longer wrote data, plus any padding necessary to
+    /// maintain derived alignment requirements.
+    pub copy_end_offset_padded: usize,
+}
+
+impl From<ComputedOffsets> for CopyRecord {
+    fn from(
+        ComputedOffsets {
+            start,
+            end,
+            end_padded,
+        }: ComputedOffsets,
+    ) -> Self {
+        Self {
+            copy_start_offset: start,
+            copy_end_offset: end,
+            copy_end_offset_padded: end_padded,
+        }
+    }
+}
+
+/// Copies `src` into the memory represented by `dst` starting at *exactly*
+/// `start_offset` bytes past the start of `dst`
+///
+/// - `start_offset` is the offset into the allocation represented by `dst`, in bytes,
+/// where the first byte of the copied data will be placed. If the requested
+/// start offset does not satisfy computed alignment requirements, an error will
+/// be returned and no data will be copied.
+///
+/// # Safety
+///
+/// This function is safe on its own, however it is very possible to do unsafe
+/// things if you read the copied data in the wrong way. See the
+/// [crate-level Safety documentation][`crate#safety`] for more.
+#[inline]
+pub fn copy_to_offset_exact<T: Copy, S: Slab>(
+    src: &T,
+    dst: &mut S,
+    start_offset: usize,
+) -> Result<CopyRecord, Error> {
+    copy_to_offset_with_align_exact(src, dst, start_offset, 1)
+}
+
+/// Copies `src` into the memory represented by `dst` starting at *exactly*
+/// `start_offset` bytes past the start of `dst` and with minimum alignment
+/// `min_alignment`.
+///
+/// - `start_offset` is the offset into the allocation represented by `dst`, in bytes,
+/// where the first byte of the copied data will be placed. If the requested
+/// start offset does not satisfy computed alignment requirements, an error will
+/// be returned and no data will be copied.
+/// - `min_alignment` is the minimum alignment to which the copy will be aligned. The
+/// copy may not actually be aligned to `min_alignment` depending on the alignment requirements
+/// of `T` (the actual alignment will be the greater between `align_of::<T>` and `min_align.next_power_of_two()`).
+///
+/// # Safety
+///
+/// This function is safe on its own, however it is very possible to do unsafe
+/// things if you read the copied data in the wrong way. See the
+/// [crate-level Safety documentation][`crate#safety`] for more.
+#[inline]
+pub fn copy_to_offset_with_align_exact<T: Copy, S: Slab>(
+    src: &T,
+    dst: &mut S,
+    start_offset: usize,
+    min_alignment: usize,
```

Comment:
Not sure I interpret this correctly:

1. It states that it copies at *exactly* `start_offset`;
2. `min_alignment` is the minimum alignment to which the copy will be aligned;
3. The copy may not actually be aligned to `min_alignment` depending on the alignment requirements of `T`.

Which is it?

Reply:
All of these things are true -- it computes the actual alignment by the greater of the two between `align_of::<T>()` and `min_alignment` and ensures that `start_offset` is indeed aligned to that computed alignment or else it will fail. The non-exact variant would instead shift the start offset up to align it up to the computed alignment requirement.

Reply:
Redid the docs for this